src/handlers/TelegramHandler.ts

```diff
@@ -0,0 +1,146 @@
+    const finalUrl = new URL("https://oauth.telegram.org/auth");
+    const clonedParams = JSON.parse(JSON.stringify(this.params.jwtParams || {}));
+    clonedParams.origin = `${this.params.redirect_uri}?state=${this.state}&nonce=${this.nonce}`;
+
+    const finalJwtParams = deepmerge(
+      {
+        state: this.state,
+        response_type: this.RESPONSE_TYPE,
+        bot_id: this.params.clientId,
+        prompt: this.PROMPT,
+        redirect_uri: `${this.params.redirect_uri}?state=${this.state}&nonce=${this.nonce}`,
+        scope: this.SCOPE,
+        nonce: this.nonce,
+      },
+      clonedParams
+    );
+    Object.keys(finalJwtParams).forEach((key: string) => {
+      const localKey = key as keyof typeof finalJwtParams;
+      if (finalJwtParams[localKey]) finalUrl.searchParams.append(localKey, finalJwtParams[localKey]);
+    });
+    this.finalURL = finalUrl;
+  }
+
+  objectToAuthDataMap(tgAuthenticationResulr: string) {
+    return JSON.parse(atob(tgAuthenticationResulr)) as { first_name: string; last_name: string; photo_url: string; username: string };
+  }
+
+  async getUserInfo(params: LoginWindowResponse): Promise<TorusVerifierResponse> {
+    const { tgAuthResult } = params;
+    const userInfo = this.objectToAuthDataMap(tgAuthResult);
+    const { photo_url: profileImage = "", username = "", first_name = "", last_name = "" } = userInfo;
+    return {
+      email: "", // Telegram does not provide email
+      name: `${first_name} ${last_name}`,
+      profileImage,
+      verifier: this.params.verifier,
+      verifierId: username.toLowerCase(),
+      typeOfLogin: this.params.typeOfLogin,
+    };
+  }
+
+  async handleLoginWindow(params: { locationReplaceOnRedirect?: boolean; popupFeatures?: string }): Promise<LoginWindowResponse> {
+    const verifierWindow = new PopupHandler({ url: this.finalURL, features: params.popupFeatures, timeout: getTimeout(this.params.typeOfLogin) });
+    if (this.params.uxMode === UX_MODE.REDIRECT) {
+      verifierWindow.redirect(params.locationReplaceOnRedirect);
+    } else {
+      const { BroadcastChannel } = await import("@toruslabs/broadcast-channel");
+      return new Promise<LoginWindowResponse>((resolve, reject) => {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        let bc: any;
+        const handleData = async (ev: string) => {
+          try {
+            const { event, origin, result, ...rest } = (JSON.parse(ev) as PopupResponse) || {};
+            // 1. Parse URL
+            const parsedUrl = new URL(origin);
+            // 2. Get state param
+            const stateParam = parsedUrl.searchParams.get("state");
+
+            if (event && event === "auth_result") {
+              if (!this.params.redirectToOpener && bc) await bc.postMessage({ success: true });
+              // properly resolve the data
+              resolve({
+                accessToken: "",
+                tgAuthResult: btoa(JSON.stringify(result)) || "",
```

Comment:
i think we shouldn't add a new field here, rather only send hash field as idToken to backend. I believe backend only need hash field from this entire result to verify